src/action/network.rs

```diff
@@ -299,3 +302,178 @@ where
+mod tests {
+
+    use super::*;
+    use std::net::{TcpStream, TcpListener, UdpSocket, SocketAddr};
+
+    /// Returns all the responses whose `local_addr` is equal to `addr`.
+    fn find_responses_by_addr<'a>(
+        session: &'a session::test::Fake,
+        addr: SocketAddr,
+    ) -> Vec<&'a Response> {
+        let mut responses = Vec::new();
+        for i in 0..session.reply_count() {
+            let resp: &Response = session.reply(i);
+            let matches = match &resp.socket_info {
+                ProtocolSocketInfo::Tcp(tcp) => {
+                    SocketAddr::new(tcp.local_addr, tcp.local_port) == addr
+                },
+                ProtocolSocketInfo::Udp(udp) => {
+                    SocketAddr::new(udp.local_addr, udp.local_port) == addr
+                },
+            };
+            if matches {
+                responses.push(resp);
+            }
+        }
+        responses
+    }
+
+    /// Returns the state of TCP connection from `response`.
+    ///
+    /// This function panics if `response` doesn't represent a TCP connection.
+    fn get_tcp_state(response: &Response) -> TcpState {
+        match &response.socket_info {
+            ProtocolSocketInfo::Tcp(tcp) => tcp.state,
+            ProtocolSocketInfo::Udp(_) => {
+                panic!("unexpected UDP socket in the response")
+            },
+        }
+    }
+
+    /// Checks if `response` represents a UDP connection.
+    fn is_udp(response: &Response) -> bool {
+        match &response.socket_info {
+            ProtocolSocketInfo::Udp(_) => true,
+            _ => false,
+        }
+    }
+
+    #[test]
+    fn test_tcp() {
+        let server = TcpListener::bind("127.0.0.1:0").unwrap();
+        let server_addr = server.local_addr().unwrap();
+        let client = TcpStream::connect(server_addr).unwrap();
+        let client_addr = client.local_addr().unwrap();
+
+        let mut session = session::test::Fake::new();
+        let request = Request { listening_only: false };
+        assert!(handle(&mut session, request).is_ok());
+
+        let our_pid = std::process::id();
+
+        // We must get two responses for the server here: one for the
+        // listening server socket and one for the temporary connection
+        // between the client and the server
+        let server_resp = find_responses_by_addr(&session, server_addr);
+        assert_eq!(server_resp.len(), 2);
+
+        let listen_resp;
+        let connection_resp;
+        if get_tcp_state(&server_resp[0]) == TcpState::Listen {
+            listen_resp = server_resp[0];
+            connection_resp = server_resp[1];
+        } else {
+            assert_eq!(get_tcp_state(&server_resp[1]), TcpState::Listen);
+            listen_resp = server_resp[1];
+            connection_resp = server_resp[0];
+        }
+
+        // We don't check PID for `connection_resp`, because it is unset at
+        // least on Linux and I don't know whether it's set on other systems.
```

Comment:
The "_I don't know..._" part looks a bit awkward and later it might be hard to say who this "_I_" is. It's better to phrase like "_it's unclear whether PID is set..._" sounds less personal.